core/src/mindustry/ui/dialogs/MapsDialog.java

```diff
@@ -108,26 +118,53 @@ void setup(){
             });
         }).size(210f, 64f);
 
-
         cont.clear();
 
-        Table maps = new Table();
-        maps.marginRight(24);
+        rebuildMaps();
 
-        ScrollPane pane = new ScrollPane(maps);
+        ScrollPane pane = new ScrollPane(mapTable);
         pane.setFadeScrollBars(false);
 
+        Table search = new Table();
+        search.image(Icon.zoom);
+        searchField = search.field("", t -> {
+            searchString = t.length() > 0 ? t.toLowerCase() : null;
+            rebuildMaps();
+        }).maxTextLength(50).growX().get();
+        searchField.setMessageText("@editor.search");
+        search.button(Icon.filter, Styles.emptyi, this::showMapFilters);
+
+        cont.add(search).growX();
+        cont.row();
+        cont.add(pane).uniformX();
+        cont.row();
+        cont.add(buttons).growX();
+    }
+
+    void rebuildMaps(){
+        mapTable.clear();
+
+        mapTable.marginRight(24);
+
         int maxwidth = Math.max((int)(Core.graphics.getWidth() / Scl.scl(230)), 1);
         float mapsize = 200f;
 
         int i = 0;
-        for(Map map : Vars.maps.all()){
+        Seq<Map> copy = showAll ? Vars.maps.all().copy() : Vars.maps.customMaps().copy();
```

Comment:
This is very inefficient; unless I'm missing something, it should be faster and easier to iterate over `showAll ? Vars.maps.all() : Vars.maps.customMaps()` and `continue;` if maps don't match a criteria. A boolean flag can keep track of empty searches.

Reply:
Fixed